Listen to new blocks

sporejack opened the pull request:
# Issue
Closes #1 .

# Changes

```
$ git log --oneline develop..1-listen-blocks
5754af5 (HEAD -> 1-listen-blocks, origin/1-listen-blocks) Update relevant dependencies
ec9626d Add note for future Websockets change
ff1dc3e Avoid `println!` in favour of explicit `writeln!`
b8287b7 Update README accordingly
2eb7de6 Add documentation for `Error` type in `main`
5936ca4 Document `client` module
760f506 Allow `Client` to hand out middleware
b22a595 (Temporarily!) disable coverage checks
cadb66e Implement block listening
73147ca Add CLI definition
1ac4527 Add client type

```

sporejack committed:
Add relevant dependencies

This commit adds some relevant dependencies.

sporejack committed:
Add client type

This commit defines the `Client` type, wrapping both a signer and a
provider into one API.

sporejack committed:
Add CLI definition

This commit defines the CLI parameters accepted by the bot.

sporejack committed:
Implement block listening

This commit adds the required functionality to listen to incoming
blocks from a specified RPC endpoint.

sporejack committed:
(Temporarily!) disable coverage checks

This commit temporarily disables minimum code coverage thresholds for
the test suite as it does not yet exist.

sporejack committed:
Allow `Client` to hand out middleware

This commit adds a new function to the `Client` API allowing a `Client`
type to create a new `SignerMiddleware` instance.

sporejack committed:
Document `client` module

This commit adds relevant documentation comments to the `client`
module. Namely,

 - Module-level comment
 - Comment for the `Client` struct
 - Comments for each of the fields of the `Client` struct
 - Comments for each function defined on the `Client` struct

sporejack committed:
Add documentation for `Error` type in `main`

This commit adds documentation comments for the global error type,
`Error`.

sporejack committed:
Update README accordingly

This commit updates the README with relevant information based on new
changes.

chrdevmar commented on src/main.rs:
```diff
@@ -1,3 +1,53 @@
+            .get_block_with_txs(block_hash)
+            .await?
+            .unwrap();
+        println!("{}", serde_json::to_string(&block).expect("Invalid data"));
```
is this the same as writing to standard output? It stood out to me and I ended up finding this

https://github.com/BurntSushi/advent-of-code/issues/17

which makes a case for using `writeln!(io::stdout, "...")` instead of `println!()`

sporejack replied:
In the same way a distinction exists in C between `fprintf(stdout, "Hello, world!")` and `printf("Hello, world!")` I suppose?

In reading the linked issue, I have to admit, I didn't know `println!` would panic on error! For what it's worth, [Lighthouse](https://github.com/sigp/lighthouse/search?q=println) doesn't seem to mind? Still good to know that `println!` panics.

sporejack replied:
@ScaredIbis Actually I agree with you here.

chrdevmar replied:
yep fair enough, my assumption after looking at this code is that this will be a building block within a larger system where it would be desirable for errors to be handled in whatever is consuming this stream of blocks

sporejack replied:
Consult ff1dc3e.

sporejack committed:
Avoid `println!` in favour of explicit `writeln!`

This commit replaces `println!` with an explicit `writeln!`.

This is because `println!` panics! Many thanks to @ScaredIbis
(https://github.com/tracer-protocol/autoclaim-bot/pull/2#discussion_r825602214).

chrdevmar commented on src/main.rs:
```diff
@@ -1,3 +1,67 @@
+#[tokio::main]
+async fn main() -> Result<(), Error> {
+    let opts: Opts = Opts::parse();
+    let client: Arc<Client<LocalWallet, Http>> = Arc::new(opts.try_into()?);
```
I think this will need to change when websocket rpc support is added but that's probs a different pr

sporejack replied:
Yeah I considered handling this for Websockets ahead of time but I think this allows more rapid local testing for now. Ideally we'd match on the provided RPC url and defer to the relevant `TryFrom` implementation in the `cli` module.

I'll add a comment to this effect.

sporejack replied:
Consult ec9626d.

sporejack committed:
Add note for future Websockets change

This commit adds a note for client parsing in the bot's entrypoint.

sporejack committed:
Update relevant dependencies

This commit updates the `syn` crate.
